src/handlers/psp.ts

```diff
@@ -256,156 +248,117 @@ const initializeProviders = async (state: State): Promise<State> => {
-      chainSubscriptions,
-      apiValuesBySubscriptionId,
-      contracts['DapiServer'],
-      voidSigner,
+    // Skip processing for the current sponsorAddress if the wallet functions fail
+    if (isNil(walletData)) {
+      const sponsorLogOptions = buildLogOptions('additional', { sponsor: sponsorAddress }, providerLogOptions);
+      utils.logger.warn('Failed to fetch sponsor wallet or transaction count', sponsorLogOptions);
+      return;
+    }
+
+    const sponsorWallet = walletData.sponsorWallet;
+
+    const sponsorWalletLogOptions = buildLogOptions(
+      'additional',
+      { sponsorWallet: shortenAddress(sponsorWallet.address) },
       providerLogOptions
     );
+    utils.logger.logPending(transactionCountLogs, sponsorWalletLogOptions);
+
+    // Get subscriptions for the current sponsorAddress and assign nonces
+    const subscriptionsWithNonce = subscriptions.map((subscription, idx) => ({
+      ...subscription,
+      nonce: walletData.transactionCount + idx,
+    }));
 
-    // Group subscriptions by sponsor address
-    const subscriptionsBySponsor = groupBy(validSubscriptions, 'sponsor');
+    utils.logger.info(`Processing ${subscriptions.length} subscription(s)`, sponsorWalletLogOptions);
 
-    // Fetch sponsor wallet transaction counts to be able to assign nonces
-    // to subscriptions and group subscriptions by sponsor wallet
-    const subscriptionsBySponsorWallets = await groupSubscriptionsBySponsorWallet(
-      subscriptionsBySponsor,
+    const logs = await processSponsorWallet(
       airnodeWallet,
-      provider,
-      currentBlock,
-      providerLogOptions
+      contracts['DapiServer'],
+      gasTarget,
+      subscriptionsWithNonce,
+      sponsorWallet,
+      voidSigner
     );
 
-    // Process sponsor wallets in parallel
-    const sponsorWalletPromises = subscriptionsBySponsorWallets.map(async ({ subscriptions, sponsorWallet }) => {
-      const sponsorWalletLogOptions = buildLogOptions(
+    logs.forEach(([logs, data]) => {
+      const subscriptionLogOptions = buildLogOptions(
         'additional',
-        { sponsorWallet: shortenAddress(sponsorWallet.address) },
-        providerLogOptions
+        { subscriptionId: data.id },
+        sponsorWalletLogOptions
       );
+      utils.logger.logPending(logs, subscriptionLogOptions);
+    });
+  });
 
-      utils.logger.info(`Processing ${subscriptions.length} subscription(s)`, sponsorWalletLogOptions);
+  await Promise.all(providerPromises);
+};
 
-      const logs = await processSponsorWallet(
-        airnodeWallet,
-        contracts['DapiServer'],
-        gasTarget,
-        subscriptions,
-        sponsorWallet
+const submitTransactions = async (state: State) => {
+  const { baseLogOptions, groupedSubscriptions, apiValuesBySubscriptionId, providerStates } = state;
+
+  const subscriptions = groupedSubscriptions.flatMap((s) => s.subscriptions);
+
+  const providerSubscriptions = providerStates.reduce(
```

Comment:
```suggestion
  const providerSponsorSubscriptions = providerStates.reduce(
```
Calling it just `providerSubscriptions` sounds misleading to me because you can have multiple objects with the same provider but different sponsor. That's actually the whole point, grouping it by provider + sponsor pair.